### ISSUE TITLE
Inconsistent lowering_config stops vectorization for elementwise + pack dispatch

### ISSUE THREAD

**Issue by hanhanW** [iree-org/iree#15574]
We are generating inconsistent lowering_config for `generic + pack` dispatch. It stops vectorization because the tile sizes in lowering_config is smaller than actual static size. This leads to a stack buffer allocation.

To repro, run `iree-opt --pass-pipeline='builtin.module(hal.executable(hal.executable.variant(iree-codegen-materialize-user-configs, iree-llvmcpu-select-lowering-strategy, iree-llvmcpu-lower-executable-target)))' ~/repro.mlir`

```mlir
hal.executable private @pack_i4 {
  hal.executable.variant public @embedded_elf_x86_64 target(<"llvm-cpu", "embedded-elf-x86_64", {cpu_features = "", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 16 : index, target_triple = "x86_64-none-elf"}>) {
    hal.executable.export public @pack_i4 ordinal(0) layout(#hal.pipeline.layout<push_constants = 0, sets = [<0, bindings = [<0, storage_buffer, ReadOnly>, <1, storage_buffer>]>]>) {
    ^bb0(%arg0: !hal.device, %arg1: index, %arg2: index):
      %x, %y, %z = flow.dispatch.workgroup_count_from_dag_root %arg1, %arg2
      hal.return %x, %y, %z : index, index, index
    }
    builtin.module {
      func.func @pack_i4() {
        %c24576 = arith.constant 24576 : index
        %c1_i4 = arith.constant 1 : i4
        %c0_i4 = arith.constant 0 : i4
        %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c24576) : !flow.dispatch.tensor<writeonly:tensor<32x344x16x32x8xi4>>
        %1 = tensor.empty() : tensor<32x344x16x32x8xi4>
        %2 = tensor.empty() : tensor<32x128x11008xi4>
        %3 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d1, d2, d0)>], iterator_types = ["parallel", "parallel", "parallel"]} outs(%2 : tensor<32x128x11008xi4>) {
        ^bb0(%out: i4):
          linalg.yield %c1_i4 : i4
        } -> tensor<32x128x11008xi4>
        %pack = tensor.pack %3 padding_value(%c0_i4 : i4) outer_dims_perm = [0, 2, 1] inner_dims_pos = [2, 1] inner_tiles = [32, 8] into %1 : tensor<32x128x11008xi4> -> tensor<32x344x16x32x8xi4>
        flow.dispatch.tensor.store %pack, %0, offsets = [0, 0, 0, 0, 0], sizes = [32, 344, 16, 32, 8], strides = [1, 1, 1, 1, 1] : tensor<32x344x16x32x8xi4> -> !flow.dispatch.tensor<writeonly:tensor<32x344x16x32x8xi4>>
        return
      }
    }
  }
}
```

If you pass `-debug-only=linalg-vectorization`, you will see:

```
[linalg-vectorization] Attempting to vectorize:
%12 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d1, d2, d0)>], iterator_types = ["parallel", "parallel", "parallel"]} outs(%11 : tensor<1x8x32xi4>) attrs =  {lowering_config = #iree_codegen.lowering_config<tile_sizes = [[64, 32, 64],
 [1, 8, 32], [0, 0, 0], [0, 0, 0]]>} {
^bb0(%out: i4):
  linalg.yield %c1_i4 : i4
} -> tensor<1x8x32xi4>

[linalg-vectorization] Input vector sizes:
1, 8, 32
[linalg-vectorization] Input scalable vector dims:
0, 0, 0
[linalg-vectorization] Iteration space static sizes:
32, 1, 8
[linalg-vectorization] Input vector sizes must be greater than or equal to iteration space static sizes
[linalg-vectorization] Vectorization pre-conditions failed
```

The lowering_config attached on `linalg.generic` is `#iree_codegen.lowering_config<tile_sizes = [[64, 32, 64], [1, 8, 32], [0, 0, 0], [0, 0, 0]]>`. And the one attached on `tensor.pack` is `#iree_codegen.lowering_config<tile_sizes = [[64, 2, 4], [1, 1, 1], [0, 0, 0], [0, 0, 0]]`.

**Comment by pzread** [comment]
The generic op has the transpose affine map: `affine_map<(d0, d1, d2) -> (d1, d2, d0)>` on parallel dims, which violates the assumption that we only fuse ops with identity map between their inputs and outputs. Therefore `setLoweringConfigForComputeOps` computes the wrong tile sizes.

I'll take a look to see the proper fix

**Comment by pzread** [comment]
I think the problem is that we unconditionally fuse pack op with linalg op and ignore the affine map on the linalg op:

https://github.com/openxla/iree/blob/7774b0defd638f50bfdd2a6a7f2411a82c842283/compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp#L504-L506

IMO this increases the chance to fuse pack op but the tile size selection and later LLVMCPUTileAndFuse are not ready to fuse ops with permuted affine maps.

I feel like to unblock this issue we can disable this kind of fuse by checking the affine map. This might reduce the fusion but we can see how to move the transpose on generic op into pack op, or support fusion with permuteted affine maps later.

**Comment by pzread** [comment]
I tested #15594 to disable fusion in this case. Since it affects the fusion, I start with a linalg IR as the input:

```mlir
func.func @pack_i4() -> tensor<32x344x16x32x8xi4> {
  %c24576 = arith.constant 24576 : index
  %c1_i4 = arith.constant 1 : i4
  %c0_i4 = arith.constant 0 : i4
  %1 = tensor.empty() : tensor<32x344x16x32x8xi4>
  %2 = tensor.empty() : tensor<32x128x11008xi4>
  %3 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d1, d2, d0)>], iterator_types = ["parallel", "parallel", "parallel"]} outs(%2 : tensor<32x128x11008xi4>) {
  ^bb0(%out: i4):
    linalg.yield %c1_i4 : i4
  } -> tensor<32x128x11008xi4>
  %pack = tensor.pack %3 padding_value(%c0_i4 : i4) outer_dims_perm = [0, 2, 1] inner_dims_pos = [2, 1] inner_tiles = [32, 8] into %1 : tensor<32x128x11008xi4> -> tensor<32x344x16x32x8xi4>
  return %pack : tensor<32x344x16x32x8xi4>
}
```
And compile with:
```sh
iree-compile --iree-hal-target-backends=llvm-cpu --iree-llvmcpu-target-cpu=cascadelake --mlir-disable-threading --iree-opt-data-tiling input.mlir
```

It created two dispatches:

Generic op dispatch:
```mlir
hal.executable public @pack_i4_dispatch_0 {
  hal.executable.variant public @embedded_elf_x86_64 target(<"llvm-cpu", "embedded-elf-x86_64", {cpu = "cascadelake", cpu_features = "+cmov,+mmx,+popcnt,+sse,+sse2,+sse3,+ssse3,+sse4.1,+sse4.2,+avx,+avx2,+fma,+avx512f,+bmi,+bmi2,+aes,+pclmul,+avx512vl,+avx512bw,+avx512dq,+avx512cd,+avx512vnni,+adx,+clflushopt,+clwb,+cx16,+cx8,+crc32,+f16c,+fsgsbase,+fxsr,+invpcid,+lzcnt,+movbe,+pku,+prfchw,+rdrnd,+rdseed,+sahf,+x87,+xsave,+xsavec,+xsaveopt,+xsaves,+evex512", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128", native_vector_size = 64 : index, target_triple = "x86_64-unknown-unknown-eabi-elf", ukernels = "default"}>) {
    hal.executable.export public @pack_i4_dispatch_0_generic_45088768_i4 ordinal(0) layout(#hal.pipeline.layout<push_constants = 0, sets = [<0, bindings = [<0, storage_buffer>]>]>) {
    ^bb0(%arg0: !hal.device):
      %x, %y, %z = flow.dispatch.workgroup_count_from_slice 
      hal.return %x, %y, %z : index, index, index
    }
    builtin.module {
      func.func @pack_i4_dispatch_0_generic_45088768_i4() {
        %c0 = arith.constant 0 : index
        %c1_i4 = arith.constant 1 : i4
        %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) : !flow.dispatch.tensor<writeonly:tensor<45088768xi4>>
        %1 = tensor.empty() : tensor<45088768xi4>
        %2 = linalg.generic {indexing_maps = [affine_map<(d0) -> (d0)>], iterator_types = ["parallel"]} outs(%1 : tensor<45088768xi4>) {
        ^bb0(%out: i4):
          linalg.yield %c1_i4 : i4
        } -> tensor<45088768xi4>
        flow.dispatch.tensor.store %2, %0, offsets = [0], sizes = [45088768], strides = [1] : tensor<45088768xi4> -> !flow.dispatch.tensor<writeonly:tensor<45088768xi4>>
        return
      }
    }
  }
}
```

Pack op dispatch:
```mlir
hal.executable public @pack_i4_dispatch_1 {
  hal.executable.variant public @embedded_elf_x86_64 target(<"llvm-cpu", "embedded-elf-x86_64", {cpu = "cascadelake", cpu_features = "+cmov,+mmx,+popcnt,+sse,+sse2,+sse3,+ssse3,+sse4.1,+sse4.2,+avx,+avx2,+fma,+avx512f,+bmi,+bmi2,+aes,+pclmul,+avx512vl,+avx512bw,+avx512dq,+avx512cd,+avx512vnni,+adx,+clflushopt,+clwb,+cx16,+cx8,+crc32,+f16c,+fsgsbase,+fxsr,+invpcid,+lzcnt,+movbe,+pku,+prfchw,+rdrnd,+rdseed,+sahf,+x87,+xsave,+xsavec,+xsaveopt,+xsaves,+evex512", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128", native_vector_size = 64 : index, target_triple = "x86_64-unknown-unknown-eabi-elf", ukernels = "default"}>) {
    hal.executable.export public @pack_i4_dispatch_1 ordinal(0) layout(#hal.pipeline.layout<push_constants = 0, sets = [<0, bindings = [<0, storage_buffer, ReadOnly>, <1, storage_buffer>]>]>) {
    ^bb0(%arg0: !hal.device):
      %x, %y, %z = flow.dispatch.workgroup_count_from_slice 
      hal.return %x, %y, %z : index, index, index
    }
    builtin.module {
      func.func @pack_i4_dispatch_1() {
        %c0 = arith.constant 0 : index
        %c0_i4 = arith.constant 0 : i4
        %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<32x128x11008xi4>>
        %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c0) : !flow.dispatch.tensor<writeonly:tensor<32x344x16x32x8xi4>>
        %2 = flow.dispatch.tensor.load %0, offsets = [0, 0, 0], sizes = [32, 128, 11008], strides = [1, 1, 1] : !flow.dispatch.tensor<readonly:tensor<32x128x11008xi4>> -> tensor<32x128x11008xi4>
        %3 = tensor.empty() : tensor<32x344x16x32x8xi4>
        %pack = tensor.pack %2 padding_value(%c0_i4 : i4) outer_dims_perm = [0, 2, 1] inner_dims_pos = [2, 1] inner_tiles = [32, 8] into %3 : tensor<32x128x11008xi4> -> tensor<32x344x16x32x8xi4>
        flow.dispatch.tensor.store %pack, %1, offsets = [0, 0, 0, 0, 0], sizes = [32, 344, 16, 32, 8], strides = [1, 1, 1, 1, 1] : tensor<32x344x16x32x8xi4> -> !flow.dispatch.tensor<writeonly:tensor<32x344x16x32x8xi4>>
        return
      }
    }
  }
}
```

When compiling the generic op dispatch, I ran into the assertion failure:
```
iree-compile: /usr/local/google/home/cheyuw/iree/iree2-build/llvm-project/tools/mlir/include/mlir/Interfaces/ViewLikeInterface.h.inc:312: int64_t mlir::detail::OffsetSizeAnd
StrideOpInterfaceTrait<mlir::memref::SubViewOp>::getStaticOffset(unsigned int) [ConcreteOp = mlir::memref::SubViewOp]: Assertion `!(*static_cast<ConcreteOp *>(this)).isDynam
icOffset(idx) && "expected static offset"' failed.                                                                                                                           
Please report issues to https://github.com/openxla/iree/issues and include the crash backtrace.                                                                              
Stack dump:                                                                                                                                                                  
0.      Program arguments: ../tools/iree-compile --iree-hal-target-backends=llvm-cpu --compile-mode=hal-executable --mlir-print-ir-before-all ./module_pack_i4_dispatch_0.mli
r                                                                                                                                                                            
 #0 0x00007f96b1a29697 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /usr/local/google/home/cheyuw/iree/iree2/third_party/llvm-project/llvm/lib/Support/Unix/Signals.in
c:723:13                                                                                                                                                                     
 #1 0x00007f96b1a278a0 llvm::sys::RunSignalHandlers() /usr/local/google/home/cheyuw/iree/iree2/third_party/llvm-project/llvm/lib/Support/Signals.cpp:106:18                  
 #2 0x00007f96b1a29d5a SignalHandler(int) /usr/local/google/home/cheyuw/iree/iree2/third_party/llvm-project/llvm/lib/Support/Unix/Signals.inc:413:1                          
 #3 0x00007f96ac05a510 (/lib/x86_64-linux-gnu/libc.so.6+0x3c510)                                                                                                             
 #4 0x00007f96ac0a80fc __pthread_kill_implementation ./nptl/./nptl/pthread_kill.c:44:76                                                                                      
 #5 0x00007f96ac05a472 raise ./signal/../sysdeps/posix/raise.c:27:6                                                                                                          
 #6 0x00007f96ac0444b2 abort ./stdlib/./stdlib/abort.c:81:7                                                                                                                  
 #7 0x00007f96ac0443d5 _nl_load_domain ./intl/./intl/loadmsgcat.c:1177:9                                                                                                     
 #8 0x00007f96ac0533a2 (/lib/x86_64-linux-gnu/libc.so.6+0x353a2)                                                                                                             
 #9 0x00007f96b40b00f5 (/usr/local/google/home/cheyuw/iree/iree2-build/lib/libIREECompiler.so+0x7ab00f5)                                                                     
#10 0x00007f96b35d076e mlir::OpConversionPattern<mlir::memref::SubViewOp>::matchAndRewrite(mlir::Operation*, llvm::ArrayRef<mlir::Value>, mlir::ConversionPatternRewriter&) c
onst /usr/local/google/home/cheyuw/iree/iree2/third_party/llvm-project/mlir/include/mlir/Transforms/DialectConversion.h:537:3                                                
#11 0x00007f96b4cf19e2 mlir::ConversionPattern::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&) const /usr/local/google/home/cheyuw/iree/iree2/third_party/llvm-pro
ject/mlir/lib/Transforms/Utils/DialectConversion.cpp:1677:10
```

The last pass IR dump is before `EmulateNarrowType`, so there might be an issue with it: 
```
// -----// IR Dump Before EmulateNarrowType (iree-codegen-emulate-narrow-type) //----- //                                                                                    
module {                                                                                                                                                                     
  func.func @pack_i4_dispatch_0_generic_45088768_i4() {                                                                                                                      
    %cst = arith.constant dense<1> : vector<16xi4>                                                                                                                           
    %c16 = arith.constant 16 : index                                                                                                                                         
    %c4096 = arith.constant 4096 : index                                                                                                                                     
    %c0 = arith.constant 0 : index                                                                                                                                           
    %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) : memref<45088768xi4>                                                
    memref.assume_alignment %0, 64 : memref<45088768xi4>                                                                                                                     
    %workgroup_id_x = hal.interface.workgroup.id[0] : index                                                                                                                  
    %1 = affine.apply affine_map<()[s0] -> (s0 * 4096)>()[%workgroup_id_x]                                                                                                   
    %subview = memref.subview %0[%1] [4096] [1] : memref<45088768xi4> to memref<4096xi4, strided<[1], offset: ?>>                                                            
    cf.br ^bb1(%c0 : index)                                                                                                                                                  
  ^bb1(%2: index):  // 2 preds: ^bb0, ^bb2                                                                                                                                   
    %3 = arith.cmpi slt, %2, %c4096 : index                                                                                                                                  
    cf.cond_br %3, ^bb2, ^bb3                                                                                                                                                
  ^bb2:  // pred: ^bb1                                                                                                                                                       
    vector.store %cst, %subview[%2] : memref<4096xi4, strided<[1], offset: ?>>, vector<16xi4>                                                                                
    %4 = arith.addi %2, %c16 : index                                                                                                                                         
    cf.br ^bb1(%4 : index)                                                                                                                                                   
  ^bb3:  // pred: ^bb1                                                                                                                                                       
    return                                                                                                                                                                   
  }                                                                                                                                                                          
}
``` 

The error can be reproduce by compiling the dispatch `@pack_i4_dispatch_0` with:
```sh
iree-compile --iree-hal-target-backends=llvm-cpu --compile-mode=hal-executable ./pack_i4_dispatch_0.mlir
```

@hanhanW do you have any idea about that error?

**Comment by hanhanW** [comment]
The error is from subbyte emulation. @Max191 has some PR on flights. Hopefully we can run it soon.

I just found that the other approach is folding transpose op into pack ops. I filed an issue for that. Disabling the aggressive fusion sounds okay to me. I can help revisit it this week.

**Comment by pzread** [comment]
I filed another issue to track the assertion error with narrow type #15608

**Comment by pzread** [comment]
Close as we no longer fuse in this case with #15594. The remaining compilation error and optimization are tracked in https://github.com/openxla/iree/issues/15608 and #15367
